framework/include/problems/FEProblemBase.h

```diff
@@ -1157,6 +1157,7 @@ class FEProblemBase : public SubProblem, public Restartable
   virtual void cacheResidual(THREAD_ID tid) override;
   virtual void cacheResidualNeighbor(THREAD_ID tid) override;
   virtual void addCachedResidual(THREAD_ID tid) override;
+  virtual void addCachedResidualTags(THREAD_ID tid, const std::set<TagID> & tags) override;
```

Comment:
I do not understand why we need this since low-level cached residuals are already handled in Assembly. `addCachedResidual` should be enough. If it is not working, we should fix this bug instead of creating a new unnecessary  API

```
void
Assembly::addCachedResiduals()
{
  for (auto tag = beginIndex(_cached_residual_values); tag < _cached_residual_values.size(); tag++)
  {
    if (!_sys.hasVector(tag))
    {
      _cached_residual_values[tag].clear();
      _cached_residual_rows[tag].clear();
      continue;
    }
    addCachedResidual(_sys.getVector(tag), tag);
  }
}
```

Reply:
BTW, this also potentially makes the tags inconsistent.

Reply:
This function only add residual to certain tagged vectors NOT all of them. That is the whole point. The nonlinear system computeResidual function is operating on certain tags, thus requires this function.

Reply:
If you want to leave some tagged vectors alone, you can  just do `disassociateVectorFromTag`, and then corresponding vectors will be ignored.

Reply:
Not in `NonlinearSystemBase::computeResidualInternal(const std::set<TagID> & tags)` without this fix. This is evidenced by the need of a `close` function call before this fix.

Reply:
In the old way, you will basically touch all vectors no matter if they are disassociated or not.

Reply:
We do not touch the disassociated vectors.

Again, you should not use `computeResidualInternal`